src/services/PostEventProcessor.ts

```diff
@@ -0,0 +1,126 @@
+import { Logger } from "@aws-lambda-powertools/logger";
+import { Metrics } from "@aws-lambda-powertools/metrics";
+
+import { AppError } from "../utils/AppError";
+import { HttpCodesEnum } from "../models/enums/HttpCodesEnum";
+import { Constants } from "../utils/Constants";
+import { IPRService } from "./IPRService";
+import { EnvironmentVariables } from "./EnvironmentVariables";
+import { ServicesEnum } from "../models/enums/ServicesEnum";
+import { createDynamoDbClient } from "../utils/DynamoDBFactory";
+import { IpvStartedOnEvent, JourneyWentAsyncOnEvent, ReadyToResumeOnEvent, AccountDeletedOnEvent, SQSEvent } from "../models/IPREventTypes";
+
+
+export class PostEventProcessor {
+	private static instance: PostEventProcessor;
+
+	private readonly logger: Logger;
+
+	private readonly metrics: Metrics;
+
+	private readonly environmentVariables: EnvironmentVariables;
+
+	private readonly iprService: IPRService;
+
+	constructor(logger: Logger, metrics: Metrics) {
+		this.logger = logger;
+		this.metrics = metrics;
+		this.environmentVariables = new EnvironmentVariables(logger, ServicesEnum.POST_EVENT_SERVICE);
+		this.iprService = IPRService.getInstance(this.environmentVariables.sessionEventsTable(), this.logger, createDynamoDbClient());
+	}
+
+	static getInstance(logger: Logger, metrics: Metrics): PostEventProcessor {
+		if (!PostEventProcessor.instance) {
+			PostEventProcessor.instance = new PostEventProcessor(logger, metrics);
+		}
+		return PostEventProcessor.instance;
+	}
+
+	async processRequest(eventBody: any): Promise<any> {
+		try {
+			const eventDetails: SQSEvent = JSON.parse(eventBody);
+			const eventName = eventDetails.event_name;
+			if (!eventDetails.user) { 
+				this.logger.error({ message: "Missing user details in event", eventDetails });
+				throw new AppError(HttpCodesEnum.SERVER_ERROR, "Missing info in sqs event");
+			}
+			const userDetails = eventDetails.user;
+
+			if (!(userDetails.user_id || userDetails.sub) || !eventDetails.timestamp) { 
+				this.logger.error({ message: "Missing required fields in event payload", eventDetails, userDetails });
+				throw new AppError(HttpCodesEnum.SERVER_ERROR, "Missing info in sqs event");
+			}
+
+			let userId = userDetails.user_id;
+
+			let updateExpression, expressionAttributeValues;
+
+			switch (eventName) {
+				case Constants.AUTH_IPV_AUTHORISATION_REQUESTED: {
```

Comment:
these events need to verify fields and be idempotent as same events with same or different fields can come through the TxMA SQS as discussed yesterday

Reply:
Verify fields logic has now been done